_posts/2017-11-14-gradle-lazy-property-evaluation.md

```diff
@@ -0,0 +1,155 @@
+
+Recently I wrote a task to configure a docker build for different Java modules. Some of them are packaged as JAR and some as WAR artifacts.
+Now this configuration was not that complicated, but I really hate duplicating stuff. So I thought how could I write a generic configuration and let each modules override some parts of this config? Thats where lazy property evaluation comes in very handy.
+
+Let's check this simple project configuration:
+```groovy
+allprojects {
+    version = '1.0.0'
+
+    ext {
+        artifactExt = "jar"
+        dockerArtifact = "${name}-${version}.${artifactExt}"
+    }
+}
+
+subprojects {
+    task printArtifactName {
+        doLast {
+            logger.lifecycle "Artifact  ${dockerArtifact}"
+        }
+    }
+}
+
+project('A') {
+    // using default configuration
+}
+
+project('B') {
+    artifactExt = 'war'
+}
+
+```
+
+The above code should do exactly what we want:
+```
+./gradlew printArtifactName
+:A:printArtifactName
+Artifact  A-1.0.0.jar
+:B:printArtifactName
+Artifact  B-1.0.0.jar
+```
+
+Wait, didn't we override the default `artifactExt` property within module B? Gradle seems to ignore the overridden property!
+
+Let's modify the example task to get a deeper insight:
+
+```groovy
+task printArtifactName {
+    doLast {
+        logger.lifecycle dockerArtifact
+        logger.lifecycle artifactExt
+    }
+}
+```
+
+```
+./gradlew printArtifactName
+:A:printArtifactName
+Artifact  A-1.0.0.jar
+Extension jar
+:B:printArtifactName
+Artifact  B-1.0.0.jar
+Extension war
+```
+
+Looks like the property `artifactExt` gets overridden correctly. The problem is caused by the evaluation time of the property `dockerArtifact`. Within Gradles configuration phase `dockerArtifact` gets evaluated directly, but at that time `artifactExt` is defined with it's default value `jar`. Later when configuring project B, `dockerArtifact` is already set and overriding `artifactExt` does not affect the value of `dockerArtifact` anymore. So we have to tell Gradle to evaluate the property `artifactExt` at execution time.
+
+We can do that by turning the property into a `Closure` like that:
+```groovy
+dockerArtifact = "${name}-${version}.${-> artifactExt}"
+```
+
+Now Gradle evaluates `name` and `version` properties eager but `artifactExt` gets evaluated lazy **each time** `dockerArtifact` is used (lazy).
+Running the modified code again gives us the expected result:
+```
+./gradlew printArtifactName
+:A:printArtifactName
+Artifact  A-1.0.0.jar
+Extension jar
+:B:printArtifactName
+Artifact  B-1.0.0.war
+Extension war
+```
+
+This simple hack can come in quite handy, but can only be used within Groovy Strings, as it uses Groovys build in [Lazy String Evaluation](http://docs.groovy-lang.org/latest/html/documentation/#_special_case_of_interpolating_closure_expressions).
+
+
+Using Closures you can also use lazy evaluation for other property types like shown below.
+
+Let's define another property called `maxMemory` as a Closure. 
```

Comment:
I would add another headline here like "Lazy Evaluation of non-String properties"